In [ ]:
import pandas as pd 
import numpy as np
import difflib as diff 
import ast

In [ ]:
df = pd.read_csv("preprocessed_train.csv",keep_default_na=False)
del df['Unnamed: 0']
df.head()

textID                                               text  \
0  cb774db0d1                I`d have responded, if I were going   
1  549e992a42      Sooo SAD I will miss you here in San Diego!!!   
2  088c60f138                          my boss is bullying me...   
3  9642c003ef                     what interview! leave me alone   
4  358bd9e861   Sons of ****, why couldn`t they put them on t...   

                         selected_text sentiment  \
0  I`d have responded, if I were going   neutral   
1                             Sooo SAD  negative   
2                          bullying me  negative   
3                       leave me alone  negative   
4                        Sons of ****,  negative   

                                          text_split  \
0  ['id', 'have', 'responded', 'if', 'i', 'were',...   
1  ['soon', 'sad', 'i', 'will', 'miss', 'you', 'h...   
2             ['my', 'boss', 'is', 'bullying', 'me']   
3      ['what', 'interview', 'leave', 'me', 'alone']   
4  ['sons', 'of', '<curse>', 'why', 'couldnt', 't...   

                                 selected_text_split initial_indice  
0  ['id', 'have', 'responded', 'if', 'i', 'were',...                 
1                                    ['soon', 'sad']                 
2                                 ['bullying', 'me']                 
3                           ['leave', 'me', 'alone']                 
4                          ['sons', 'of', '<curse>']

The next step in this process is to setup the start and end indices that the model will be trained on. We can set this up using either character indices or word indices,we need to decide which. It doesnt make a ton of sense to predit indices that cut a word in half so doing this based on characters probably wont work

As we moved onto finding the indi``ces we noticed a new error: Our split text columns are not lists, they are strings made to look like a list. When we saved to csv for some reason it converted the lists to strings. We are going to use a library called ast to fix this issue. We plan to literally evaluate each string as a list. The library will read in the string i.e "[1,2,4]" and convert it to [1,2,3] which is usable viable python code. Note: This is usually a terrible terrible idea and we acknowledge this. It is however the simplest solution to the issue we are having and as this is a low stakes evironment primarily for learning we accept the usage of literal_eval. For more information read the following [link](https://stackoverflow.com/questions/1832940/why-is-using-eval-a-bad-practice)

In [ ]:
df["text_split"] = df.text_split.apply(lambda x: ast.literal_eval(x))
df["selected_text_split"] = df.selected_text_split.apply(lambda x: ast.literal_eval(x))
df.head()

textID                                               text  \
0  cb774db0d1                I`d have responded, if I were going   
1  549e992a42      Sooo SAD I will miss you here in San Diego!!!   
2  088c60f138                          my boss is bullying me...   
3  9642c003ef                     what interview! leave me alone   
4  358bd9e861   Sons of ****, why couldn`t they put them on t...   

                         selected_text sentiment  \
0  I`d have responded, if I were going   neutral   
1                             Sooo SAD  negative   
2                          bullying me  negative   
3                       leave me alone  negative   
4                        Sons of ****,  negative   

                                          text_split  \
0          [id, have, responded, if, i, were, going]   
1  [soon, sad, i, will, miss, you, here, in, san,...   
2                       [my, boss, is, bullying, me]   
3                [what, interview, leave, me, alone]   
4  [sons, of, <curse>, why, couldnt, they, put, t...   

                         selected_text_split initial_indice  
0  [id, have, responded, if, i, were, going]                 
1                                [soon, sad]                 
2                             [bullying, me]                 
3                         [leave, me, alone]                 
4                        [sons, of, <curse>]

Unsurprisingly this data is pretty dirty, so there are plenty of instances where the selected text is really different from the main text. To deal with this we have told get_close_matches to have no cutoff, just return the closest match regardless of similarity.

In [ ]:
def index_finder(text,selected,index):
    text = list(text)
    selected = list(selected)
    return text.index(diff.get_close_matches(selected[index],text,cutoff=0)[0])
df["initial_indice"] = df.apply(lambda x: index_finder(x.text_split,x.selected_text_split,0),axis=1)
df["final_indice"] = df.apply(lambda x: index_finder(x.text_split,x.selected_text_split,-1),axis=1)
df.head()

textID                                               text  \
0  cb774db0d1                I`d have responded, if I were going   
1  549e992a42      Sooo SAD I will miss you here in San Diego!!!   
2  088c60f138                          my boss is bullying me...   
3  9642c003ef                     what interview! leave me alone   
4  358bd9e861   Sons of ****, why couldn`t they put them on t...   

                         selected_text sentiment  \
0  I`d have responded, if I were going   neutral   
1                             Sooo SAD  negative   
2                          bullying me  negative   
3                       leave me alone  negative   
4                        Sons of ****,  negative   

                                          text_split  \
0          [id, have, responded, if, i, were, going]   
1  [soon, sad, i, will, miss, you, here, in, san,...   
2                       [my, boss, is, bullying, me]   
3                [what, interview, leave, me, alone]   
4  [sons, of, <curse>, why, couldnt, they, put, t...   

                         selected_text_split  initial_indice  final_indice  
0  [id, have, responded, if, i, were, going]               0             6  
1                                [soon, sad]               0             1  
2                             [bullying, me]               3             4  
3                         [leave, me, alone]               2             4  
4                        [sons, of, <curse>]               0             2

In [ ]:
df[df.initial_indice > df.final_indice]

textID                                               text  \
134    6903cb08f2   Nice to see you tweeting!  It`s Sunday 10th M...   
166    c78bf59e67    #lichfield #tweetup sounds like fun  Hope to...   
321    e5160216a5    ohh my tooth is hurts ohh  im sad it very hurts   
957    c400a6bf99   alas, I am moving   (like where i`m moving to...   
999    6d2170e60f   I`ve been unlocked for decades now...just not...   
...           ...                                                ...   
26618  fd4e4be626   Michelle,  is on here. Don`t u miss her being...   
26927  5ba1ad2348    Morning!  Happy Mother`s Day Miz!!!  Thank y...   
26953  528ee63138   He was my reason for even going too haha.  Sh...   
27433  705ca50b20  At least they ( kids) havnt gotten to big to s...   
27476  4eac33d1c0   wish we could come see u on Denver  husband l...   

                                           selected_text sentiment  \
134                                               e nice  positive   
166                                    p sounds like fun  positive   
321                                 im sad it very hurts  negative   
957                               h) wish I could go too  positive   
999                    .just not lucky, never have been.  negative   
...                                                  ...       ...   
26618                ? I liked your interviews with her.  positive   
26927                                            !  Happ  positive   
26953                                             c haha  positive   
27433  AWWW they are so sweet. I am so lucky to have ...  positive   
27476                                             d lost  negative   

                                              text_split  \
134    [nice, to, see, you, sweeting, its, sunday, <n...   
166    [lichfield, #tweetup, sounds, like, fun, hope,...   
321    [ohh, my, tooth, is, hurts, ohh, im, sad, it, ...   
957    [alas, i, am, moving, like, where, ibm, moving...   
999    [ive, been, unlocked, for, decades, now, just,...   
...                                                  ...   
26618  [michelle, is, on, here, dont, u, miss, her, b...   
26927  [morning, happy, mothers, day, miz!!!, thank, ...   
26953  [he, was, my, reason, for, even, going, too, a...   
27433  [at, least, they, (, kids, haunt, gotten, to, ...   
27476  [wish, we, could, come, see, u, on, denver, hu...   

                                     selected_text_split  initial_indice  \
134                                            [e, nice]              17   
166                               [p, sounds, like, fun]               5   
321                           [im, sad, it, very, hurts]               6   
957                        [he, wish, i, could, go, too]              10   
999                [just, not, lucky, never, have, been]               6   
...                                                  ...             ...   
26618         [?, i, liked, your, interviews, with, her]              13   
26927                                          [!, happ]               4   
26953                                           [c, aha]              15   
27433  [aww, they, are, so, sweet, i, am, so, lucky, ...              15   
27476                                          [d, lost]              12   

       final_indice  
134               0  
166               4  
321               4  
957               8  
999               1  
...             ...  
26618             7  
26927             1  
26953             8  
27433             4  
27476             9  

[293 rows x 8 columns]

The issue here is that in places where the selected text is particularly messy the intial index is actually after the selected index. This is a big issue as we absolutely do not want to be telling our model that it is ok to label and initial index as a final index. Since there were only 293 times where that happens which is relatively few we will choose to drop those samples from the dataset 

In [ ]:
df = df[df.initial_indice < df.final_indice]
df[df.initial_indice > df.final_indice]

Empty DataFrame
Columns: [textID, text, selected_text, sentiment, text_split, selected_text_split, initial_indice, final_indice]
Index: []

Now there are none that violate that condition!

We can now move to tokenization and preparing the code for modeling!

In [ ]:
df.to_csv("feature_engineered.csv")